src/Microsoft.Azure.WebJobs.Host/Executors/TriggeredFunctionExecutor.cs

```diff
@@ -51,15 +55,41 @@ public async Task<FunctionResult> TryExecuteAsync(TriggeredFunctionData input, C
                 };
             }
 
+            // initial attempt
             IFunctionInstance instance = _instanceFactory.Create(context);
-
             IDelayedException exception = await _executor.TryExecuteAsync(instance, cancellationToken);
 
+            // Retry
+            if (_descriptor.RetryStrategy != null && exception != null)
+            {
+                exception = await TryExecuteWithRetriesAsync(context, cancellationToken);
+            }
+
             FunctionResult result = exception != null ?
                 new FunctionResult(exception.Exception)
                 : new FunctionResult(true);
 
             return result;
         }
+
+        internal async Task<IDelayedException> TryExecuteWithRetriesAsync(FunctionInstanceFactoryContext<TTriggerValue> context, CancellationToken cancellationToken)
```

Comment:
To remove this code duplication, I was thinking that we could have a single static extension method for IFunctionExecutor with signature like `TryExecuteAsync(Func<IFunctionInstance> instanceFactory, ILogger logger, CancellationToken)`. You can then call this same method in all places needed as a one liner replacement.

Note - there's also one more place where we need to do this: https://github.com/Azure/azure-webjobs-sdk/blob/b798412ad74ba97cf2d85487ae8479f277bdd85c/src/Microsoft.Azure.WebJobs.Host/Executors/HostMessageExecutor.cs#L127. Now this path isn't really used in v3, but we should fix it up for completeness. It's for Dashboard integration, when someone hits "Replay" in the dashboard, and an internal queue message is used to trigger the invocation.